src/README.md

```diff
@@ -0,0 +1,858 @@
+[`Environment`][] instance. The restriction of using a single event loop
+is not inherent to the design of Node.js, and a sufficiently committed person
+could restructure Node.js to provide e.g. the ability to run parts of Node.js
+inside an event loop separate from the active thread’s event loop.
+
+<a id="environment"></a>
+### `Environment`
+
+Node.js instances are represented by the `Environment` class.
+
+Currently, every `Environment` class is associated with:
+
+* One [event loop][]
+* One [`Isolate`][]
+* One main [`Context`][]
+
+The `Environment` class contains a large number of different fields for
+different Node.js modules, for example a libuv timer for `setTimeout()` or
+the memory for a `Float64Array` that the `fs` module uses for storing data
+returned from a `fs.stat()` call.
+
+It also provides [cleanup hooks][] and maintains a list of [`BaseObject`][]
+instances.
+
+Typical ways of accessing the current `Environment` in the Node.js code are:
+
+* Given a `FunctionCallbackInfo` for a [binding function][],
+  using `Environment::GetCurrent(args)`.
+* Given a [`BaseObject`][], using `env()` or `self->env()`.
+* Given a [`Context`][], using `Environment::GetCurrent(context)`.
+  This requires that `context` has been associated with the `Environment`
+  instance, e.g. is the main `Context` for the `Environment` or one of its
+  `vm.Context`s.
+* Given an [`Isolate`][], using `Environment::GetCurrent(isolate)`. This looks
+  up the current [`Context`][] and then uses that.
+
+<a id="isolate-data"></a>
+### `IsolateData`
+
+Every Node.js instance ([`Environment`][]) is associated with one `IsolateData`
+instance that contains information about or associated with a given
+[`Isolate`][].
+
+#### String table
+
+`IsolateData` contains a list of strings that can be quickly accessed
+inside Node.js code, e.g. given an `Environment` instance `env` the JavaScript
+string “name” can be accessed through `env->name_string()` without actually
+creating a new JavaScript string.
+
+### Platform
+
+Every process that uses V8 has a `v8::Platform` instance that provides some
+functionalities to V8, most importantly the ability to schedule work on
+background threads.
+
+Node.js provides a `NodePlatform` class that implements the `v8::Platform`
+interface and uses libuv for providing background threading abilities.
+
+The platform can be accessed through `isolate_data->platform()` given an
+[`IsolateData`][] instance, although that only works when:
+
+* The current Node.js instance was not started by an embedder; or
+* The current Node.js instance was started by an embedder whose `v8::Platform`
+  implementation also implement’s the `node::MultiIsolatePlatform` interface
+  and who passed this to Node.js.
+
+<a id="binding-functions"></a>
+### Binding functions
+
+C++ functions exposed to JS follow a specific signature. The following example
+is from `node_util.cc`:
+
+```c++
+void ArrayBufferViewHasBuffer(const FunctionCallbackInfo<Value>& args) {
```

Comment:
Roughly starting from here I think it becomes a bit confusing which types mentioned here are provided by V8 and which are provided by Node.js.  It'd be good to always include the namespaces, or explain that namespaces are omitted in the code examples.

Reply:
I’ve added a sentence here saying that we usually use `using` statements to avoid typing out namespaces.

I think for examples taken from our source code it makes sense to keep the code as-is as far as it makes sense, because after all this is intended to document how we actually write code inside Node.js…